Folded LightCurve fix

Nschanche opened the pull request.

Nschanche commented:
This WIP PR will address Issues #1429 and #1422. These issues arise because normalized phase folded light curves are dimensionless, and therefore functions like astropy's aggregate_downsample do not work.

Nschanche commented:
- [ ] Figure out wrap_phase #1423
- [ ] Add tests for this new functionality
- [ ] Check if this fixes #1271

orionlee commented:
It seems like the PR proposes a breaking behavior change on `lc.fold()` in general, such that `lc.time` is no longer the phase (be it in TimeDelta or normalized phase).

If that's the case, 
- I feel it should be done in a major version upgrade (say in v3), as opposed to a minor point release.
- if the issue should be resolved at upstream astorpy TimeSeries.
  I just double checked with the current Astropy 7.1 dev codes. The issue remains there. For a normalized folded TimeSeries, with time of `Quantity`: 
  a. `bin()`: underlying `aggregate_downsample()` raises AttributeError
  b. `copy()`: TimeSeries `copy()` raises ValueError

orionlee commented:
An alternative approach that could fix the issues without any behavior change or upstream astropy change.

For each affected method, if it's a FoldedLightCurve with normalized phase
- create a temporary copy and maps normalized phase in Quantity to TimeDelta (say, in JD, the unit does not matter)
- perform the operation on the temporary copy.
  - time in TimeDelta type is the workaround to make the operations work (that assumes time is of `Time` type)
- convert the result's time column from the TimeDelta back to Quantity.

An outline of such approach: 

```python
class FoldedLightcurve:
    # ...

    def _to_tmp_lc_with_fake_time(self):
        # create a copy with  time column (normalized phase)
        # changed from Quantity to a TimeDelta
        pass

    def _restore_normalized_phase_to_time(self):
        # create a copy: change time column (logically normalized phase)
        # from the temporary TimeDelta back to Quantity
        pass

    def select_flux(self, flux_column, flux_err_column=None):
        lc = self
        if isinstance(self.time, Quantity):  # case in normalized phase
            lc = self._to_tmp_lc_with_fake_time()  # now the time (normalized phase) is of type TimeDelta
        lc = lc.select_flux(flux_column, flux_err_column)
        lc = lc._restore_normalized_phase_to_time()
        return lc
```

Nschanche commented:
Thanks @orionlee, I like your second suggestion. It functionally does the same thing, but is less intrusive and less likely to cause problems for users. I will pivot to that structure and keep pushing on this WIP.

orionlee commented:
It's probably obvious but I'd state it as a reminder: if we follow my suggestion above, the override for `lc.bin()` needs to explicitly raise errors for binning with time based parameters (`time_bin_size`, etc.) for the normalized phase case.

orionlee commented:
Nschanche, I've reported the problem in upstream astropy TimeSeries: 
https://github.com/astropy/astropy/issues/18070

Nschanche commented:
@orionlee can you review the PR and let me know your thoughts? It may be confusing as normalized phase folded light curves are 'unnormalized' (reverted back to TimeDelta time series) before binning, so the time steps are required to be in time units. I didn't add an explicit warning about this, just left it in the doctoring. I also restricted inputs to the FoldedTimeSeries binning function to be either time units or number of bins.

orionlee commented:
@Nschanche I'll take a look by early next week.

orionlee commented on src/lightkurve/lightcurve.py:
```diff
@@ -3140,7 +3195,113 @@ def plot_river(self, **kwargs):
+    def copy(self):
+
+    
+        if self.normalize_phase == True:
```
Consider to replace the implementation with something like this:
```python
    def copy(self):
        # the default copy() in astropy fails if the lightcurve
        # has normalized phase (time column is `Quantity` instead of `Time` like)
        # workaround it by temporarily changing time column to 
        # non-normalized `TimeDelta`
        try:
            if self.normalize_phase == True:
                self._replace_normalized_phase()
            result = super().copy()
        finally:
            if self.normalize_phase == True:
                self._normalize_the_phase()
                result._normalize_the_phase()

        return result
```

1. simplify `result._delay_required_column_checks():` block with `result._normalize_the_phase()` ( or `result._restore_normalized_phase()` if it's renamed )
2. Wrap around with a try finally block so that if something unexpectedly go wrong during actual copy(), `self` will be restored back to the original state.
3. The comment is also applicable to the similar code path in `bin()`.

Nschanche replied:
I updated the simplification suggested in 1 for this and for bin()

orionlee commented on src/lightkurve/lightcurve.py:
```diff
@@ -3140,7 +3195,113 @@ def plot_river(self, **kwargs):
+        time_bin_size=None,
+        time_bin_start=None,
+        aggregate_func=None,
+        bins=None,
```
- Shouldn't the parameters here be the same as the `bin()` of regular (non-folded) lightcurve?
  - shouldn't `n_bins` (and probably `binsize`) be supported as well?
  - if `time_bin_size` is supported even for a folded lightcurve with normalized phase, shouldn't `time_bin_end` be included as well?

Nschanche replied:
Actually, looking into time_bin_end more closely, it looks like it doesn't really get passed in the the regular bin() function. I do want to remove it as an option (I will add that in the change file).

orionlee replied:
Hmm, it does look `time_bin_end` parameter never gets passed in regular bin(), an issue dated from the introduction of the feature in #1042 (v2.1.0). 

I'm confused about what you intend to do:
1. On the one hand, for regular lightcurve, codes are added in this PR to properly pass along `time_bin_end` to the underlying `aggregate_downsample()`.
2. On the other hand, for FoldedLightCurve, `time_bin_end` is added as a parameter but is not passed along.

Nschanche replied:
I made the last commit before finishing my edits. I intend to remove time_bin_end as an optional keyword for FoldedLightCurves:
    def bin(
        self,
        time_bin_size=None,
        time_bin_start=None,
        aggregate_func=None,
        bins=None,
        n_bins=None,
    ):

orionlee commented on src/lightkurve/lightcurve.py:
```diff
@@ -3247,8 +3408,9 @@ def to_fits(
         for kw in kepler_specific_data:
-            if ~np.asarray([kw.lower == k.lower() for k in extra_data]).any():
-                extra_data[kw] = kepler_specific_data[kw]
+            if hasattr(self, kw):
```
This change in `to_fit()` appears to be unrelated. Some kind of merge problems?

Nschanche replied:
I don't know why this change ended up here. I will revert.

Nschanche commented on tests/test_lightcurve.py:
```diff
@@ -257,7 +257,7 @@ def test_lightcurve_fold():
     assert_array_equal(np.sort(fold.time_original), lc.time)
     assert len(fold.time_original) == len(lc.time)
     fold = lc.fold(period=1, epoch_time=-0.1)
-    assert_almost_equal(fold.time[0], -0.5, 2)
```
Great, phase is an alias for time, so time should also be valid. I've added this test back in.

Nschanche commented on tests/test_lightcurve.py:
```diff
@@ -280,6 +280,14 @@ def test_lightcurve_fold():
     with pytest.warns(LightkurveWarning, match="appears to be given in JD"):
         lc.fold(10, 2456600)
 
+    # Make sure normalizing and binning work for folded lightcurves
```
Updated comment with "Make sure binning a phase-normalized folded lightcurve works (#1422)"

Nschanche commented on tests/test_lightcurve.py:
```diff
@@ -280,6 +280,14 @@ def test_lightcurve_fold():
+    assert len(fold.bin(bins=10)) == 10
+    fold = lc.fold(period=1.5, normalize_phase=True)
+    assert_almost_equal(np.max(fold.phase)-np.min(fold.phase), 1, 1)
+    assert len(fold.bin(bins=10)) == 10
```
Added
`    
fold_copy = fold.copy()
    assert_array_equal(fold.time, fold_copy.time)
    assert_array_equal(fold.flux, fold_copy.flux)
   `

Nschanche commented on src/lightkurve/lightcurve.py:
```diff
@@ -3054,6 +3072,39 @@ def even_mask(self):
+
+        # If the lightcurve is phase folded but not normalized, just return self
+        if self.normalize_phase == False:
+            return self
```
Great, I've updated this

Nschanche commented on src/lightkurve/lightcurve.py:
```diff
@@ -3054,6 +3072,39 @@ def even_mask(self):
+            warnings.warn("Not a phase-folded lightcurve, no phase available")
+
+        # If the lightcurve is phase folded but not normalized, just return self
+        if self.normalize_phase == False:
```
OK, this looks like a good warning to me. I've replaced it

Nschanche commented on src/lightkurve/lightcurve.py:
```diff
@@ -3054,6 +3072,39 @@ def even_mask(self):
+        # Some astropy functions, such as aggregate_downsample, require Time or TimeDelta
+        # As normalized phase-folded lightcurves are unitless, this breaks
+        # This will re-normalized the phase
+        with self._delay_required_column_checks():
```
Check added

orionlee commented on src/lightkurve/lightcurve.py:
```diff
@@ -3140,7 +3206,125 @@ def plot_river(self, **kwargs):
+            else:
+                warnings.warn("The function cannot be run on a lightcurve with normalized phase.")
+
+        if time_bin_size != None:
```
For `time_bin_size` semantics in the case of normalized phase folded LC, I suggest the following changes to make it more intuitive to users.

```python

# lc_f is a folded lightcurve with normalized phase with 10 day period

# supported by the PR. No change
lc_fb = lc_f.bin(time_bin_size=1.0*u.day)  

# not supported by the PR
# suggestion: support it, 
#  internally converts it from normalized phase to time (with lc_f.period * time_bin_size)
lc_fb = lc_f.bin(time_bin_size=0.1 * u.dimensionless_unscaled)

# PR treats it the same as 0.1 * u.day, the behavior of generic LightCurve.fold()
# it's probably surprising to users
# suggestion: treat it as 0.1 * u.dimensionless_unscaled  (in normalized phase)
lc_fb = lc_f.bin(time_bin_size=0.1)  
```

The implementation can be something like this:
```python
        if time_bin_size is not None:
            if not isinstance(time_bin_size, Quantity):
                # case time_bin_size is a plain float / int
                if self.normalize_phase:
                    time_bin_size = time_bin_size * u.dimensionless_unscaled
                else:
                    time_bin_size = time_bin_size * u.day  # duplicating the logic in parent .bin()
            if self.normalize_phase and time_bin_size.unit == u.dimensionless_unscaled:
                # aggregate_downsample() does not work on normalized phase
                # convert time_bin_size from normalized phase to time
                time_bin_size = self.period * time_bin_size.value
```

orionlee commented on src/lightkurve/lightcurve.py:
```diff
@@ -3140,7 +3206,125 @@ def plot_river(self, **kwargs):
+        if time_bin_start is None:
+            time_bin_start = self.time[0]
+        if np.isscalar(time_bin_size):
+            if isinstance(self.time, TimeDelta):
```
The `TimeDelta` check is not necessary, as `self.time` has been temporarily converted to `TimeDelta` (if needed) by the earlier line `self._replace_normalized_phase()`

Nschanche replied:
Good catch

orionlee commented on CHANGES.rst:
```diff
@@ -11,6 +11,7 @@
 - Updated the default TESS quality mask to include Argabrightening, Impulsive Outlier, and Bad Calibration Exclude flags (see issue #1230). [#1336]
 - Updated tutorial notebook links [#1484]
 - Author keyword now specifies between SPOC and TESS-SPOC [#1487]
+- Modified copy() and bin() functions for FoldedLightCurve objects [#1491]
```
1. Reword it to state the purpose (of fixing them for case FoldedLightCurve with normalized phase)
2. If `time_bin_end`,  `n_bins`, and `binsize` will no longer be supported for `FoldedLightCurve`, state it here as well given it is a breaking change.

Nschanche replied:
Added an additional comment for these deprecated keywords.

orionlee reviewed (changes requested):
See the new comments and the conversations I unresolved. The main things:

1. concerns of breaking changes by dropping `n_bins`, `time_bin_end` and `binsize` for FoldedLightCurve *without* normalized phase (i.e., time in `TimeDelta`)
2. make `time_bin_size` more intuitive for FoldedLightCurve with normalized phase: supporting `u.dimensionless_unscaled`
3. Tests for `lc.copy()` remain to be missing.

Nschanche commented on tests/test_lightcurve.py:
```diff
@@ -280,6 +281,19 @@ def test_lightcurve_fold():
+    binned = fold.bin(bins=10)
+    assert len(binned) == 10
+    # ensure fold was not changed
+    assert len(fold) == 100
```
Gotcha, I've added this assertion.

Nschanche commented on src/lightkurve/lightcurve.py:
```diff
@@ -3054,6 +3074,48 @@ def even_mask(self):
+        # Some astropy functions, such as aggregate_downsample, require Time or TimeDelta
+        # As normalized phase-folded lightcurves are unitless, this breaks
+        # This will replace the normalized phase with phase in TimeDelta
+        if not hasattr(self, 'normalize_phase'):
```
I think I somehow didn't commit the copy tests, which were quite simple
    fold_copy = fold.copy()
    assert_array_equal(fold.time, fold_copy.time)
    assert_array_equal(fold.flux, fold_copy.flux)

orionlee commented on src/lightkurve/lightcurve.py:
```diff
@@ -3140,7 +3204,143 @@ def plot_river(self, **kwargs):
 
+        Parameters
+        ----------
+        time_bin_size : `~astropy.units.Quantity` or `~astropy.time.TimeDelta`, optional
```
The docstring for time_bin_size should be updated to explain the semantics for normalized phase (that it can accept either time or normalized phase).

Nschanche commented on tests/test_lightcurve.py:
```diff
@@ -287,12 +287,22 @@ def test_lightcurve_fold():
     assert len(fold) == 100
+    assert isinstance(fold.time, u.Quantity)
+
+    # Make sure 'copy()' works as expected
```
I have added this additional assertions.

Nschanche commented:
@orionlee this test keeps timing out, but I think this addresses your concerns. I want to merge this asap as we want to release a new lightkurve version before the new TIKE image is released next week. Any final thought on this (or other PRs you strongly think should be imminently merged)?

orionlee commented:
I think this PR is okay to be merged. It'd be even better if change log mentions fixing the `bin(bin=###)` case ( issue #1271, it is for all FoldedLightCurve, not just those with normalized phase). Something like:

```rst
- Fixed the bug in ``bin()`` function with ``bin` parameter for ``FoldedLightCurve`` objects. [#1491]
```

---

I'd love to have PR #1472 merged, but it's not a showstopper.

orionlee commented:
Also, a note about github issue management: Once this PR is merged, you probably need to manually close all the issues.

If you change the wording in the comment to something like "Fixes hashtag<the issue number>", github will automatically link the issues and close them once the PR is merge. See:  https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue

orionlee commented on src/lightkurve/lightcurve.py:
```diff
@@ -3140,7 +3204,141 @@ def plot_river(self, **kwargs):
+ 
+
+        if bins != None:
+            # astropy's aggregate_downsample doesn't work for phase data:
```
I just realized the comment `astropy's aggregate_downsample doesn't work for phase data` is incorrect.  The problem of `bin(bins=###)` is due to lightkurve's own `Lighcurve.bin()` codes that assume time column is of `Time` type.

https://github.com/lightkurve/lightkurve/blob/b81ebc207a5cad3040659fd141f75fa7136a678f/src/lightkurve/lightcurve.py#L1543-L1544

It'd be better to reword it to something like "parent bin() does not work for phase data"
